addons/sale_margin/models/sale_order.py

```diff
@@ -70,5 +73,19 @@ class SaleOrder(models.Model):
 
     @api.depends('order_line.margin')
     def _product_margin(self):
-        for order in self:
-            order.margin = sum(order.order_line.filtered(lambda r: r.state != 'cancel').mapped('margin'))
+        if self.env.in_onchange:
+            for order in self:
+                order.margin = sum(order.order_line.filtered(lambda r: r.state != 'cancel').mapped('margin'))
+        else:
+            # On batch records recomputation (e.g. at install), compute the margins
+            # with a single read_group query for better performance.
+            # This isn't done in an onchange environment because (part of) the data
+            # may not be stored in database (new records or unsaved modifications).
+            grouped_order_lines_data = self.env['sale.order.line'].read_group(
+                [
+                    ('order_id', 'in', self.ids),
+                    ('state', '!=', 'cancel'),
+                ], ['margin', 'order_id'], ['order_id'])
+            for data in grouped_order_lines_data:
```

Comment:
generally we write like this : (and it write on each record in self (compatible for V13)).

```python
data = dict((d['order_id'][0], d['margin'])
                    for d in self.env['sale.order.line'].read_group(
                [('order_id', 'in', self.ids), ('state', '!=', 'cancel'),
                ], ['margin', 'order_id'], ['order_id']))
for order in self:
     order.margin = data.get(order.id, 0)
```

Reply:
@nim-odoo

Reply:
cc @mart-e